nvflare/app_common/abstract/statistics_spec.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) 2021-2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict, List
+
+from nvflare.apis.fl_component import FLComponent
+from nvflare.apis.fl_context import FLContext
+from nvflare.app_common.statistics.stats_def import Feature, Histogram
+
+
+class Statistics(FLComponent):
+    def initialize(self, parts: dict, fl_ctx: FLContext):
+        """Initialize the Statistics generator object. This is called before the Statistics can perform calculation
+
+        This is called only once.
+
+        Args:
+            parts: components to be used by the Statistics generator
+            fl_ctx: FLContext of the running environment
+        """
+        pass
+
+    def features(self) -> Dict[str, List[Feature]]:
+        """
+           return feature names for each dataset.
+           For example, we have training and test datasets.
+
+           the method will return
+           { "train": features1, "test": features2}
+           where features1,2 are the list of Features with contains feature name and DataType
+
+        :return: Dict[<dataset_name>, List[Feature]]
+        """
```

Comment:
Please change all docstrings to Google style: https://google.github.io/styleguide/pyguide.html

Reply:
The method description should go right after triple "

For example:

```
def fetch_smalltable_rows(table_handle: smalltable.Table,
                          keys: Sequence[Union[bytes, str]],
                          require_all_keys: bool = False,
) -> Mapping[bytes, tuple[str, ...]]:
    """Fetches rows from a Smalltable.

    Retrieves rows pertaining to the given keys from the Table instance
    represented by table_handle.  String keys will be UTF-8 encoded.

    Args:
        table_handle: An open smalltable.Table instance.
        keys: A sequence of strings representing the key of each table
          row to fetch.  String keys will be UTF-8 encoded.
        require_all_keys: If True only rows with values set for all keys will be
          returned.

    Returns:
        A dict mapping keys to the corresponding table row data
        fetched. Each row is represented as a tuple of strings. For
        example:

        {b'Serak': ('Rigel VII', 'Preparer'),
         b'Zim': ('Irk', 'Invader'),
         b'Lrrr': ('Omicron Persei 8', 'Emperor')}

        Returned keys are always bytes.  If a key from the keys argument is
        missing from the dictionary, then that row was not found in the
        table (and require_all_keys must have been False).

    Raises:
        IOError: An error occurred accessing the smalltable.
    """
```